swift-sdk/IterableInboxViewController.swift

```diff
@@ -129,92 +118,84 @@ open class IterableInboxViewController: UITableViewController {
             cell.iconImageView?.isHidden = true
         }
         
-        if let createdAt = viewModel.createdAt {
+        if let createdAt = message.createdAt {
             cell.createdAtLbl?.isHidden = false
             cell.createdAtLbl?.text = IterableInboxViewController.displayValue(forTime: createdAt)
         } else {
             cell.createdAtLbl?.isHidden = true
         }
     }
+    
+    private static func displayValue(forTime date: Date) -> String {
+        return DateFormatter.localizedString(from: date, dateStyle: .medium, timeStyle: .short)
+    }
+}
 
-    private func loadImage(forMessageId messageId: String, fromUrl url: URL) {
-        if let networkSession = IterableAPI.internalImplementation?.networkSession {
-            NetworkHelper.getData(fromUrl: url, usingSession: networkSession).onSuccess {[weak self] in
-                self?.updateCell(forMessageId: messageId, withImageData: $0)
-            }.onError {
-                ITBError($0.localizedDescription)
-            }
-        }
+extension IterableInboxViewController : InboxViewControllerViewModelDelegate {
+    // Must be called on Main Thread
+    func onViewModelChanged(diff: [SectionedDiffStep<Int, InboxMessageViewModel>]) {
+        ITBInfo()
+
+        updateTableView(diff: diff)
+        updateUnreadBadgeCount()
     }
-    
-    private func updateCell(forMessageId messageId: String, withImageData data: Data) {
-        guard var viewModels = diffCalculator?.rows else {
-            return
-        }
-        guard let row = viewModels.firstIndex (where: { $0.iterableMessage.messageId == messageId }) else {
-            return
-        }
-        let viewModel = viewModels[row]
-        viewModel.imageData = data
 
+    // Must be called on Main Thread
+    func onImageLoaded(forRow row: Int) {
+        ITBInfo()
         tableView.reloadRows(at: [IndexPath(row: row, section: 0)], with: .automatic)
     }
+
+    private func updateUnreadBadgeCount() {
+        let unreadCount = viewModel.unreadCount
+        let badgeValue = unreadCount == 0 ? nil : "\(unreadCount)"
+        navigationController?.tabBarItem?.badgeValue = badgeValue
+    }
     
-    private static func displayValue(forTime date: Date) -> String {
-        return DateFormatter.localizedString(from: date, dateStyle: .medium, timeStyle: .short)
+    private func updateTableView(diff: [SectionedDiffStep<Int, InboxMessageViewModel>]) {
+        tableView.beginUpdates()
+        viewModel.beganUpdates()
+        for result in diff {
+            switch result {
+            case let .delete(section, row, _): tableView.deleteRows(at: [IndexPath(row: row, section: section)], with: self.deletionAnimation)
+            case let .insert(section, row, _): tableView.insertRows(at: [IndexPath(row: row, section: section)], with: self.insertionAnimation)
+            case let .sectionDelete(section, _): tableView.deleteSections(IndexSet(integer: section), with: self.deletionAnimation)
```

Comment:
IMO: It seems like using the Dwifft library might be overkill since we never update the sections. Do you envision using that functionality in the future? 

It looks like you are trying to keep a list of InboxMessageViewModel and add/remove entries based on which ones are changed from the server. Since we are guaranteed that only new items are added at the end of the new list, then we can we can iterate through the list and compare if the items match to figure out the difference.

Reply:
Possible solution you could use to get the corresponding delete/inserts. You might have to modify it slightly depending on if it changes the row count after calling deleteRows.

`let oldList = ["Car", "Skate", "Bike", "Plane", "Boat"]
let newList = ["Car","Bike","Boat", "Camel","Kite"]

var newListCounter = 0
for (index, element) in oldList.enumerated() {
    let item = newList[newListCounter]
    if (item != element) {
        print("delete: \(index), \(element)")
    } else {
        newListCounter+=1
    }
}

let itemsToAdd = newList[newListCounter...]
print("Items to add \(itemsToAdd)")`

Reply:
Unfortunately it is not going to be as straightforward as this. There are going to be updates and deletes also. Let us keep it like it is for now and replace it with the built-in new way when the next version of Xcode is released.